horovod/common/operations.cc

```diff
@@ -995,57 +1023,72 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
         OP_ERROR(entries, ex.what())
       }
       DDL_CHECK(entries, "ddl_allreduce",
-                ddl_allreduce(buffer_data,
-                              (size_t)num_elements,
-                              ddl_data_type,
+                ddl_allreduce(buffer_data, (size_t)num_elements, ddl_data_type,
                               DDL_OP_SUM))
 #else
       if (horovod_global.hierarchical_allreduce) {
-        NCCL_CHECK(entries, "ncclReduce",
-                   ncclReduce(fused_input_data, buffer_data,
-                              (size_t)num_elements,
-                              GetNCCLDataType(first_entry.tensor), ncclSum, 0,
-                              nccl_comm, stream))
+        int64_t num_elements_per_gpu = num_elements / horovod_global.local_size;
+
+        size_t buffer_len_per_proc =
+            (buffer_len / num_elements) * num_elements_per_gpu;
+
+        void* buffer_data_at_gpu_offset =
+            (uint8_t*)buffer_data +
+            buffer_len_per_proc * horovod_global.local_rank;
+
+        NCCL_CHECK(entries, "ncclReduceScatter",
+                   ncclReduceScatter(fused_input_data,
+                                     buffer_data_at_gpu_offset,
+                                     (size_t)num_elements_per_gpu,
+                                     GetNCCLDataType(first_entry.tensor),
+                                     ncclSum, nccl_comm, stream))
+
         if (timeline.Initialized()) {
-          RECORD_EVENT(entries, event_queue, NCCL_REDUCE, stream)
+          RECORD_EVENT(entries, event_queue, NCCL_REDUCESCATTER, stream)
         }
 
-        if (horovod_global.local_rank == 0) {
-          // cudaHostAlloc is significantly slower than malloc.  Pre-allocating
-          // a buffer is not safe since the tensor can be arbitrarily large.
-          host_buffer = malloc(buffer_len);
+        // cudaHostAlloc is significantly slower than malloc.  Pre-allocating
+        // a buffer is not safe since the tensor can be arbitrarily large.
 
-          CUDA_CHECK(entries, "cudaMemcpyAsync",
-                     cudaMemcpyAsync(host_buffer, buffer_data, buffer_len,
-                                     cudaMemcpyDeviceToHost, stream))
-          // This event must be recorded for the subsequent synchronize.
-          RECORD_EVENT(entries, event_queue, MEMCPY_IN_HOST_BUFFER, stream)
+        host_buffer = malloc(buffer_len_per_proc);
```

Comment:
Why host_buffer needed here? MPI-cuda aware can handle gpu memory.

Reply:
CUDA-aware MPI + NCCL can deadlock each other and thus are not recommended combination (see #435).